src/util/psbt/mod.rs

```diff
@@ -210,6 +213,234 @@ impl PartiallySignedTransaction {
 
         Ok(())
     }
+
+    /// Signs the input at `input_index` with private key `sk`.
+    ///
+    /// # Errors
+    /// If an attempt is made to sign a taproot input.
+    pub fn sign<C: Signing>(&mut self, sk: &PrivateKey, input_index: usize, secp: &Secp256k1<C>) -> Result<(), SighashError> {
+        self.check_index_is_within_bounds(input_index)?;
+
+        match self.signing_algorithm(input_index)? {
+            Algo::Ecdsa => self.sign_ecdsa(sk, input_index, secp),
+            Algo::Schnorr => Err(SighashError::TaprootNotYetSupported),
+        }
+    }
+
+    fn sign_ecdsa<C: Signing>(&mut self, sk: &PrivateKey, input_index: usize, secp: &Secp256k1<C>) -> Result<(), SighashError> {
+        let mut cache = SighashCache::new(&self.unsigned_tx);
+        let (msg, sighash_ty) = self.sighash_ecdsa(input_index, &mut cache)?;
+
+        let sig = secp.sign_ecdsa(&msg, &sk.inner);
+
+        let mut final_signature = Vec::with_capacity(75);
+        final_signature.extend_from_slice(&sig.serialize_der());
+        final_signature.push(sighash_ty.to_u32() as u8);
+
+        let pk = sk.public_key(secp);
+        self.inputs[input_index]
+            .partial_sigs
+            .insert(pk, EcdsaSig::from_slice(&final_signature)?);
+
+        Ok(())
+    }
+
+    /// Returns the sighash message to sign an ECDSA input along with the sighash type.
+    fn sighash_ecdsa<T: Deref<Target=Transaction>>(
+        &self,
+        input_index: usize,
+        cache: &mut SighashCache<T>,
+    ) -> Result<(Message, PsbtSighashType), SighashError> {
+        self.check_index_is_within_bounds(input_index)?;
+
+        let input = &self.inputs[input_index];
+        let utxo = self.spend_utxo(input_index)?;
+        let script = utxo.script_pubkey.clone(); // scriptPubkey for input spend utxo.
```

Comment:
We should be able to avoid this clone here

Reply:
You are correct, thanks!